File: sdk/src/types/knowledge.ts
export interface KnowledgeDocument {
  fileName: string;
  fileSize: number;
  mimeType: string;
  chunkCount: number;
  uploadedAt: string;
}

/** Like Skill/Agent, this mirrors the real wire shape (`_id`, raw domain/ownerType fields). */
export interface KnowledgeBase {
  _id: string;
  domain: string;
  ownerType: 'PersonaUser' | 'Project' | 'ExternalUser';
  ownerId?: string;
  externalOwnerId?: string;
  name: string;
  description?: string;
  isPublic: boolean;
  documentCount: number;
  chunkCount: number;
  qdrantCollectionName: string;
  documents: KnowledgeDocument[];
  embeddingModel: string;
  providerId?: string;
  chunkSize: number;
  chunkOverlap: number;
  topK: number;
  createdAt: string;
  updatedAt: string;
}

export interface CreateKnowledgeBaseInput {
  name: string;
  description?: string;
  isPublic?: boolean;
  embeddingModel?: string;
  /**
   * Required on the Developer API — unlike the Persona route, there's no
   * "my default provider" concept for a Project or ExternalUser caller.
   */
  providerId: string;
  chunkSize?: number;
  chunkOverlap?: number;
  topK?: number;
}

export interface UpdateKnowledgeBaseInput {
  name?: string;
  description?: string;
  isPublic?: boolean;
  embeddingModel?: string;
  providerId?: string;
  chunkSize?: number;
  chunkOverlap?: number;
  topK?: number;
}

export interface DiscoverKnowledgeBasesParams {
  page?: number;
  limit?: number;
  search?: string;
  /** Restricts to the asserted external user's own Knowledge Bases (ProjectRuntimeContext only). */
  scope?: 'mine';
}

export interface UploadFileInput {
  filename: string;
  /** Node Buffer/Uint8Array, or an already-constructed Blob. */
  content: Uint8Array | Blob;
  contentType?: string;
}

export interface UploadDocumentsResult {
  documentCount: number;
  chunkCount: number;
  files: Array<{ fileName: string; fileSize: number; mimeType: string; chunkCount: number }>;
}

export interface DeleteDocumentResult {
  removedChunks: number;
  remainingDocuments: number;
  remainingChunks: number;
}

export interface KnowledgeSearchResult {
  text: string;
  source: string;
  score: number | null;
}

File: sdk-python/src/personaai/types/knowledge.py
"""Mirrors ``sdk/src/types/knowledge.ts``."""

from __future__ import annotations

from typing import BinaryIO, Literal, TypedDict


class KnowledgeDocument(TypedDict):
    fileName: str
    fileSize: int
    mimeType: str
    chunkCount: int
    uploadedAt: str


class _KnowledgeBaseRequired(TypedDict):
    _id: str
    domain: str
    ownerType: Literal["PersonaUser", "Project", "ExternalUser"]
    name: str
    isPublic: bool
    documentCount: int
    chunkCount: int
    qdrantCollectionName: str
    documents: list[KnowledgeDocument]
    embeddingModel: str
    chunkSize: int
    chunkOverlap: int
    topK: int
    createdAt: str
    updatedAt: str


class KnowledgeBase(_KnowledgeBaseRequired, total=False):
    """Like ``Skill``/``Agent``, this mirrors the real wire shape (``_id``,
    raw domain/ownerType fields)."""

    ownerId: str
    externalOwnerId: str
    description: str
    providerId: str


class _CreateKnowledgeBaseInputRequired(TypedDict):
    name: str
    # required on the Developer API — unlike the Persona route, there's no
    # "my default provider" concept for a Project or ExternalUser caller
    providerId: str


class CreateKnowledgeBaseInput(_CreateKnowledgeBaseInputRequired, total=False):
    description: str
    isPublic: bool
    embeddingModel: str
    chunkSize: int
    chunkOverlap: int
    topK: int


class UpdateKnowledgeBaseInput(TypedDict, total=False):
    name: str
    description: str
    isPublic: bool
    embeddingModel: str
    providerId: str
    chunkSize: int
    chunkOverlap: int
    topK: int


class DiscoverKnowledgeBasesParams(TypedDict, total=False):
    page: int
    limit: int
    search: str
    scope: Literal["mine"]  # restricts to the asserted external user's own KBs (runtime-only)


class _UploadFileInputRequired(TypedDict):
    filename: str
    content: bytes | BinaryIO


class UploadFileInput(_UploadFileInputRequired, total=False):
    contentType: str


class _UploadedFileSummary(TypedDict):
    fileName: str
    fileSize: int
    mimeType: str
    chunkCount: int


class UploadDocumentsResult(TypedDict):
    documentCount: int
    chunkCount: int
    files: list[_UploadedFileSummary]


class DeleteDocumentResult(TypedDict):
    removedChunks: int
    remainingDocuments: int
    remainingChunks: int


class KnowledgeSearchResult(TypedDict):
    text: str
    source: str
    score: float | None
